util/ranger/ranger.go

```diff
@@ -184,31 +184,69 @@ func convertPoint(sctx sessionctx.Context, point *point, tp *types.FieldType) (*
 	return npoint, nil
 }
 
-// appendPoints2Ranges appends additional column ranges for multi-column index.
-// The additional column ranges can only be appended to point ranges.
-// for example we have an index (a, b), if the condition is (a > 1 and b = 2)
-// then we can not build a conjunctive ranges for this index.
-func appendPoints2Ranges(sctx sessionctx.Context, origin []*Range, rangePoints []*point,
-	ft *types.FieldType) ([]*Range, error) {
-	var newIndexRanges []*Range
+func appendPoint2Range(origin *Range, startPoint, endPoint *point, ft *types.FieldType) *Range {
+	lowVal := make([]types.Datum, len(origin.LowVal)+1)
+	copy(lowVal, origin.LowVal)
+	lowVal[len(origin.LowVal)] = startPoint.value
+
+	highVal := make([]types.Datum, len(origin.HighVal)+1)
+	copy(highVal, origin.HighVal)
+	highVal[len(origin.HighVal)] = endPoint.value
+
+	collators := make([]collate.Collator, len(origin.Collators)+1)
+	copy(collators, origin.Collators)
+	collators[len(origin.Collators)] = collate.GetCollator(ft.GetCollate())
+	return &Range{
+		LowVal:      lowVal,
+		LowExclude:  startPoint.excl,
+		HighVal:     highVal,
+		HighExclude: endPoint.excl,
+		Collators:   collators,
+	}
+}
+
+// appendPoints2Ranges appends additional column ranges for multi-column index. The additional column ranges can only be
+// appended to point ranges. For example, we have an index (a, b), if the condition is (a > 1 and b = 2), then we can not
+// build a conjunctive ranges for this index.
+// rangeMaxSize is the max memory limit for ranges. O indicates no memory limit.
+// If the second return value is true, it means that the estimated memory usage of ranges after appending points exceeds
+// rangeMaxSize and the function rejects appending points to ranges.
+func appendPoints2Ranges(sctx sessionctx.Context, origin Ranges, rangePoints []*point,
+	ft *types.FieldType, rangeMaxSize int64) (Ranges, bool, error) {
+	// Estimate whether rangeMaxSize will be exceeded first before appending points to ranges.
+	if rangeMaxSize > 0 && len(origin) > 0 && len(rangePoints) > 0 {
+		startPoint, err := convertPoint(sctx, rangePoints[0], ft)
+		if err != nil {
+			return nil, false, errors.Trace(err)
+		}
+		endPoint, err := convertPoint(sctx, rangePoints[1], ft)
+		if err != nil {
+			return nil, false, errors.Trace(err)
+		}
+		ran := appendPoint2Range(origin[0], startPoint, endPoint, ft)
+		if ran.MemUsage()*int64(len(origin))*int64(len(rangePoints))/2 > rangeMaxSize {
+			return origin, true, nil
+		}
+	}
```

Comment:
What if the first rangePoint has large/tiny data which caused the estimate difference too big?